### PR TITLE
Network behaviour: Adjust TCP timeout and keepalive settings

### PR DESCRIPTION
**Pull request by amotl**
Hi there,

after reading #33, I believe @RyanW8 might have been tripped by https://github.com/crate/crate/issues/10779. On the Python driver, @chaudum mitigated that by contributing https://github.com/crate/crate-python/pull/374. We might want to do similar things here?

This patch just uses those default values:

- KeepAlive: 15 seconds
- Timeout: 10 seconds

Ideally, we could make it configurable, but nevertheless would need some reasonable default values.

What do you think?

With kind regards,
Andreas.

### CODE REVIEW

**Comment by chaudum** [comment]
Yes, makes sense to also apply useful network settings here as well.

Reducing the keepalive interval is a good idea, but why do you want to use a different default than in https://github.com/crate/crate-python/pull/374 ? I've seen Golang's default is 15s.

I'm not sure about adding a timeout, though. I understand, that neither reads nor writes should take longer than 10s, but one never knows... If we want to introduce a timeout, it should definitely be configurable IMO.

**Comment by amotl** [comment]
> Reducing the keepalive interval is a good idea, but why do you want to use a different default than in crate/crate-python#374 ? I've seen Golang's default is 15s.

Ok. What do you believe we should use there instead?

> I'm not sure about adding a timeout, though. I understand, that neither reads nor writes should take longer than 10s, but one never knows... If we want to introduce a timeout, it should definitely be configurable IMO.

I believe it will be the connect timeout, not a read/write timeout. But we can make it configurable, sure. What do you think would be a reasonable default?

**Comment by chaudum** [comment]
> > Reducing the keepalive interval is a good idea, but why do you want to use a different default than in [crate/crate-python#374](https://github.com/crate/crate-python/pull/374) ? I've seen Golang's default is 15s.
> 
> Ok. What do you believe we should use there instead? Shall we just leave out configuring this and trust that Golang's default of 15s will apply? Any chance to validate that?

_If zero, keep-alive probes are sent with a default value (currently 15 seconds), if supported by the protocol and operating system._

[dial.go:L75](https://golang.org/src/net/dial.go)

I would use the same value as in other clients (30s).

> 
> > I'm not sure about adding a timeout, though. I understand, that neither reads nor writes should take longer than 10s, but one never knows... If we want to introduce a timeout, it should definitely be configurable IMO.
> 
> I believe it will be the connect timeout, not a read/write timeout. But we can make it configurable, sure. What do you think would be a reasonable default?

You're right, it's the connect timeout:

_Timeout is the maximum amount of time a dial will wait for a connect to complete._

[dial.go:L29](https://golang.org/src/net/dial.go)

File: CHANGES.rst [hunk 1]
```diff
@@ -5,6 +5,11 @@ CHANGES for CrateDB Prometheus Adapter
 Unreleased
 ==========
 
+- Network behaviour: Adjust TCP timeout and keepalive settings
```
[the remark this answers is not in the document]
Comment (amotl) [reply CHANGES.rst]: Thank you very much!

**Pull request merged**